components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSimplePAPStatusDataHandler.java

```diff
@@ -0,0 +1,389 @@
+    }
+
+    /**
+     * Deletes all status records.
+     *
+     * @param about whether the status is about a policy or publisher.
+     * @param key   key value of the status.
+     * @throws EntitlementException if fails to delete.
+     */
+    public void deletePersistedData(String about, String key) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+        String query = EntitlementConstants.Status.ABOUT_POLICY.equals(about) ?
+                DELETE_POLICY_STATUS_SQL : DELETE_SUBSCRIBER_STATUS_SQL;
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (NamedPreparedStatement deleteStatusPrepStmt = new NamedPreparedStatement(connection, query)) {
+                deleteStatusPrepStmt.setString(KEY, key);
+                deleteStatusPrepStmt.setInt(TENANT_ID, tenantId);
+                deleteStatusPrepStmt.executeUpdate();
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException("Error while deleting policy status", e);
+        }
+    }
+
+    public void persistStatus(String about, String key, List<StatusHolder> statusHolders) throws EntitlementException {
+
+        boolean useLastStatusOnly = Boolean.parseBoolean(
+                IdentityUtil.getProperty(EntitlementConstants.PROP_USE_LAST_STATUS_ONLY));
+
+        if (statusHolders != null && !statusHolders.isEmpty()) {
+
+            if (useLastStatusOnly) {
+                // Delete the previous status
+                deletePersistedData(about, key);
+                auditAction(statusHolders.toArray(new StatusHolder[0]));
+            }
+
+            // Add new status to the database
+            addStatus(about, key, statusHolders);
+
+            if (!useLastStatusOnly) {
+                Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+                try {
+                    // Get the existing status count
+                    int statusCount = getStatusCount(connection, about, key);
+
+                    // Delete old status data if the count exceeds the maximum records
+                    if (statusCount > maxRecords) {
+                        deleteStatus(connection, about, key, statusCount);
+                    }
+                    IdentityDatabaseUtil.commitTransaction(connection);
+                } catch (SQLException e) {
+                    IdentityDatabaseUtil.rollbackTransaction(connection);
+                    throw new EntitlementException("Error while deleting surplus policy status", e);
+                } finally {
+                    IdentityDatabaseUtil.closeConnection(connection);
+                }
+            }
+        }
+    }
+
+    private List<StatusHolder> getStatus(String key, String about) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
```

Comment:
Shall we pass tenantId as a param to keep consistency of other pure DAO methods?